src/classes/TemporaryFile.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  S3Client,
+  CreateMultipartUploadCommand,
+  UploadPartCommand,
+  CompleteMultipartUploadCommand,
+} from '@aws-sdk/client-s3';
+import { v4 as uuid } from 'uuid';
+import { SystemConfigurationError } from '../errors';
+import type {
+  CompletedPart,
+} from '@aws-sdk/client-s3';
+
+export class TemporaryFile {
+  public readonly permanentFileSystemPath: string;
+
+  private uploadedLength = 0;
+
+  private s3Url?: string;
+
+  private uploadId?: string;
+
+  private readonly bucket: string;
+
+  private readonly key: string;
+
+  private partNumber = 1;
+
+  private readonly s3: S3Client;
+
+  private readonly completedParts: CompletedPart[] = [];
+
+  public constructor(permanentFileSystemPath: string) {
+    if (process.env.TEMPORARY_FILE_S3_BUCKET === undefined) {
+      throw new SystemConfigurationError('TEMPORARY_FILE_S3_BUCKET must be populated in order to upload to s3.');
+    }
+    if (process.env.AWS_ACCESS_KEY_ID === undefined) {
+      throw new SystemConfigurationError('AWS_ACCESS_KEY_ID must be populated in order to upload to s3.');
+    }
+    if (process.env.AWS_ACCESS_SECRET === undefined) {
+      throw new SystemConfigurationError('AWS_ACCESS_SECRET must be populated in order to upload to s3.');
+    }
+    if (process.env.TEMPORARY_FILE_S3_BUCKET_REGION === undefined) {
+      throw new SystemConfigurationError('TEMPORARY_FILE_S3_BUCKET_REGION must be populated in order to upload to s3.');
+    }
+    const path = `${process.env.TEMPORARY_FILE_S3_SUBDIRECTORY ?? ''}/unprocessed`;
+    this.s3 = new S3Client({
+      region: process.env.TEMPORARY_FILE_S3_BUCKET_REGION,
+      credentials: {
+        accessKeyId: process.env.AWS_ACCESS_KEY_ID,
+        secretAccessKey: process.env.AWS_ACCESS_SECRET,
+      },
+    });
+    this.bucket = process.env.TEMPORARY_FILE_S3_BUCKET;
+    this.key = `${path}/${uuid()}`;
+    this.permanentFileSystemPath = permanentFileSystemPath;
+  }
+
+  public get url(): string {
+    if (this.s3Url === undefined) {
+      throw new Error();
+    }
+    return this.s3Url;
+  }
+
+  public get size(): number {
+    return this.uploadedLength;
+  }
+
+  public async open(): Promise<void> {
+    const multipartUploadCommand = new CreateMultipartUploadCommand({
+      Bucket: this.bucket,
+      Key: this.key,
+      ContentType: 'application/octet-stream',
+    });
+    const response = await this.s3.send(multipartUploadCommand);
+    this.uploadId = response.UploadId;
+  }
+
+  public async append(fileData: Buffer): Promise<void> {
+    if (this.uploadId === undefined) {
+      throw new Error('You cannot post data before opening the file');
+    }
+    const uploadPartCommand = new UploadPartCommand({
+      Bucket: this.bucket,
+      Key: this.key,
+      Body: fileData,
+      UploadId: this.uploadId,
+      PartNumber: this.partNumber,
+    });
+    const completedPart = await this.s3.send(uploadPartCommand);
```

Comment:
I'm not sure where the right place is to check for this in this use case, but S3 won't accept a part smaller than 5 MB unless it's the last part.

Reply:
Oh dear... that's a big one!  I'm certain SFTP will be getting chunks smaller than 5 MB.  This would mean I need to write to temporary storage as well.

Thanks for the heads up; I'll need to do some refactoring.

Reply:
For context on this limit (posting for me / historians): https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html